### PR TITLE
enhancement(tls): add option to set alpn protocols in tls settings

### PR DESCRIPTION
**Pull request by tesharp**
This is a more general enhancement of problem described in #9901. Instead of setting alpn protocols just for vector source this enhancement adds an option on the TlsConfig so it can be reused for other sources / sinks if necessary. Does not change behavior as it is if not defined.

Example when running against nginx ingress to force it to only use h2:

```yaml
sinks:
  vector:
    type: vector
    inputs:
    - everything
    address: vector.sink.example.com:443
    tls:
      enabled: true
      alpn_protocols:
      - h2
    version: "2"
```

(First commit with rust, so excuse my potential mistakes..)

### CODE REVIEW

**Comment by jszwedko** [comment]
Thanks @tesharp ! We'll try to get this reviewed next week.

File: src/tls/settings.rs [hunk 1]
```diff
@@ -317,6 +333,20 @@ impl TlsConfig {
         }
     }
 
+    fn parse_alpn_protocols(&self) -> Result<Option<Vec<u8>>> {
```
Comment (neuronull) [inline src/tls/settings.rs:336]: Might be helpful to add a comment here describing the format required:

The input must be in ALPN "wire format". It consists of a sequence of supported protocol names prefixed by their byte length.

**Review by neuronull — COMMENTED** [review]
Thanks for the contribution!

I have just a couple suggestions below.

Going to request a review from someone more familiar with the TLS config area.

A couple other items:

- It looks like there is a CI failure unrelated to your changes, you will probably want to merge in master to resolve.

- Would you be able to add a couple unit test cases to the unit tests at the bottom of the file ? I think it would be good to validate a successful parsing/application if the config options to the context, as well as an invalid parsing/application to the context.

- Would you mind updating the docs for the TLS settings? I believe they are in `website/cue/reference/components.cue`. 
  This aims to describe the documentation process with cue, but it can tricky if you haven't used it before (though good job with your first Rust!).
   If not, I can take care of updating the docs for you.
   
   https://github.com/vectordotdev/vector/blob/master/docs/DOCUMENTING.md

File: src/tls/mod.rs [hunk 2]
```diff
@@ -103,6 +103,8 @@ pub enum TlsError {
     AddCertToStore { source: ErrorStack },
     #[snafu(display("Error setting up the verification certificate: {}", source))]
     SetVerifyCert { source: ErrorStack },
+    #[snafu(display("Error setting alpn protocols: {}", source))]
```
Comment (bruceg) [inline src/tls/mod.rs:106]: This should probably be in all-caps to indicate it's an acronym: "ALPN"

**Review by bruceg — DISMISSED** [review]
I second all of @neuronull's comments. I just have one tiny suggestion below.

**Comment by jszwedko** [comment]
This PR seemed like it went stale so I pushed a couple of commits addressing the feedback if you get a chance to review again @bruceg @neuronull

File: src/tls/settings.rs [hunk 3]
```diff
@@ -283,10 +283,9 @@ impl TlsSettings {
             load_mac_certs(context).unwrap();
         }
 
-        if self.alpn_protocols.is_some() {
-            let alpn = self.alpn_protocols.as_ref().unwrap();
+        if let Some(apln) = &self.alpn_protocols {
             context
-                .set_alpn_protos(alpn.as_slice())
+                .set_alpn_protos(apln.as_slice())
```
Comment (neuronull) [inline src/tls/settings.rs:288]: perhaps name the variable "alpn" to avoid confusion? Which I realize now that was my suggestion 😆

**Review by neuronull — APPROVED** [review]

File: src/tls/settings.rs (hunk 3 above)
Comment (jszwedko) [reply src/tls/settings.rs:288]: Doh! I made that misspelling many times. Thanks for catching.

**Comment by jszwedko** [comment]
Thanks for getting the ball rolling on this @tesharp !

**Pull request merged**